### PR TITLE
Get input devices

### PR DESCRIPTION
**Pull request by Spider-Min**

### CODE REVIEW

File: live/client.py [hunk 1]
```diff
@@ -23,6 +25,19 @@ class InvalidResponse(ElementalException):
     pass
 
 
+def etree_to_dict(tree_node):
```
Comment (flavioribeiro) [inline live/client.py]: I think you can use a library for that like https://pypi.org/project/xmljson/ or https://micropyramid.com/blog/how-to-convert-xml-content-into-json-using-xmltodict/

File: live/client.py [hunk 2]
```diff
@@ -132,3 +147,52 @@ def stop_event(self, event_id):
...
+
+        # Find all devices info
+        devices_info = []
+        all_devices = set()
+        for device in devices_list:
+            all_devices.add(device.find('device_name').text)
+            device_dic = etree_to_dict(device)
+            devices_info.append(device_dic['device'])
+
+        # Find in use devices from all events
```
Comment (flavioribeiro) [inline live/client.py]: can you extract this in another method? something like `find_devices_in_use()`

File: live/client.py [hunk 3]
```diff
@@ -132,3 +147,52 @@ def stop_event(self, event_id):
...
+        for d in in_use_devices:
+            devices_availability[d] = False
+
+        # Append availability info to device info
+        for device in devices_info:
+            device['availability'] = \
+                devices_availability[device['device_name']]
+
+        devices_info = sorted(devices_info, key=lambda a: int(a["id"]))
+        print(json.dumps(devices_info))
```
Comment (flavioribeiro) [inline live/client.py]: remove this print

File: live/client.py [hunk 4]
```diff
@@ -132,3 +147,52 @@ def stop_event(self, event_id):
...
+        return json.dumps(devices_info)
```
Comment (igorsobreira) [inline live/client.py]: The client method should the python object, the list of dicts.

In the controller, if we are using that in a http response, then we serialize to json, but not at this level.

File: live/client.py [hunk 5]
```diff
@@ -132,3 +147,52 @@ def stop_event(self, event_id):
...
+            devices_availability[d] = True
+        for d in in_use_devices:
+            devices_availability[d] = False
+
+        # Append availability info to device info
+        for device in devices_info:
+            device['availability'] = \
+                devices_availability[device['device_name']]
+
+        devices_info = sorted(devices_info, key=lambda a: int(a["id"]))
```
Comment (igorsobreira) [inline live/client.py]: Optional python tip: the `operator` module
```python
>>> l = [{'id': '3'}, {'id': '2'}, {'id': '1'}]
>>> sorted(l, key=lambda d: d['id'])
[{'id': '1'}, {'id': '2'}, {'id': '3'}]

>>> from operator import itemgetter
>>> sorted(l, key=itemgetter('id'))
[{'id': '1'}, {'id': '2'}, {'id': '3'}]
```

File: live/client.py [hunk 6]
```diff
@@ -132,3 +147,52 @@ def stop_event(self, event_id):
...
+
+        # Find in use devices from all events
+        in_use_devices = set()
+        for event in events_list:
+            status = event.find('status').text
+            for input in event.findall('input'):
+                device_input = input.find('device_input')
+                if device_input:
+                    device_name = device_input.find('device_name')
+                    if status == 'preprocessing' or status == 'running':
```
Comment (igorsobreira) [inline live/client.py]: Python tip: for `or`s like this, specially if you have more than one (not this case) you can use tuples:

```python
if status in ('preprocessing',  'running'):
```

File: live/client.py [hunk 7]
```diff
@@ -132,3 +134,58 @@ def stop_event(self, event_id):
...
+        events_xml = self.send_request(
+            http_method="GET", url=events_url, headers=events_headers)
+
+        events_list = ET.fromstring(events_xml.text)
+
+        # Find in use devices from all events
+        in_use_devices = set()
+        for event in events_list:
+            status = event.find('status').text
+            for input in event.findall('input'):
```
Comment (flavioribeiro) [inline live/client.py]: `input` is a reserved word for an internal method, it would be good to change this variable name

File: live/client.py [hunk 8]
```diff
@@ -132,3 +134,58 @@ def stop_event(self, event_id):
...
+    def get_input_devices(self):
+        devices_url = f'{self.server_ip}/devices'
+        devices_headers = self.generate_headers(devices_url)
+        devices_xml = self.send_request(
+            http_method="GET", url=devices_url, headers=devices_headers)
+
+        devices_list = ET.fromstring(devices_xml.text)
+
+        # Find all devices info
+        all_devices = set()
```
Comment (flavioribeiro) [inline live/client.py]: this one could be:
```python
all_devices = [device.find('device_name').text for device in devices_list]
```

File: live/client.py [hunk 9]
```diff
@@ -132,3 +134,58 @@ def stop_event(self, event_id):
...
+
+        devices_info_dict = xmltodict.parse(devices_xml.text)[
+            'device_list']['device']
+
+        in_use_devices = self.find_devices_in_use()
+
+        devices_availability = {}
+
+        for d in all_devices:
+            devices_availability[d] = True
```
Comment (flavioribeiro) [inline live/client.py]: You could also use a dictionary comprehension to  generate the `available_devices`. Something like:

```python
>>> all_devices = ["dev1", "dev2", "dev3", "dev4", "dev5"]
>>> devices_in_use = ["dev1", "dev3"]
>>> {d:(d not in devices_in_use) for d in all_devices}
{'dev3': False, 'dev2': True, 'dev1': False, 'dev5': True, 'dev4': True}
```

Comment (Spider-Min) [reply live/client.py]: cool!

File: live/client.py [hunk 10]
```diff
@@ -132,3 +134,52 @@ def stop_event(self, event_id):
...
+        devices_xml = self.send_request(
+            http_method="GET", url=devices_url, headers=devices_headers)
+
+        devices_list = ET.fromstring(devices_xml.text)
+
+        # Find all devices info
+        all_devices = \
+            {device.find('device_name').text for device in devices_list}
+
+        devices_info_dict = xmltodict.parse(devices_xml.text)[
```
Comment (flavioribeiro) [inline live/client.py]: the `_dict` in the end of  this variable is an anti-pattern: https://docs.quantifiedcode.com/python-anti-patterns/readability/putting_type_information_in_a_variable_name.html

File: live/client_test.py [hunk 11]
```diff
@@ -198,3 +198,94 @@ def test_stop_event_should_call_send_request_as_expect():
...
+    client.send_request = mock.Mock()
+    client.find_devices_in_use = mock.Mock()
+    client.find_devices_in_use.return_value = ("HD-SDI 1",)
+    client.send_request.return_value = \
+        mock_response(status=200,
+                      text=file_fixture('sample_device_list.xml'))
+
+    res = client.get_input_devices()
+    print(res)
+    assert res == [{"@href": "/devices/1", "id": "1",
```
Comment (flavioribeiro) [inline live/client_test.py]: we  don't need the  `@href` in the payload

File: live/client.py [hunk 12]
```diff
@@ -157,17 +157,14 @@ def get_input_devices(self):
...
         devices_in_use = self.find_devices_in_use()
 
-        devices_availability = \
-            {d: (d not in devices_in_use) for d in all_devices}
-
         for device in devices_info:
             device.pop('@href')
-            device['availability'] = \
-                devices_availability[device['device_name']]
+            if device['device_name'] in devices_in_use:
```
Comment (flavioribeiro) [inline live/client.py]: you can remove the `if` and just do:
```python
device['availability'] = (device['device_name'] in devices_in_use)
```

**Pull request merged**